### ISSUE TITLE
Filtering may ignore clustering key restrictions if they form a prefix without a partition key

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#4541]
Ref: https://github.com/scylladb/scylla/issues/3803

The solution for issue #3803 provided a way of fetching columns that are not included in the select clause if they are needed for filtering. However, one corner case is still unsolved: if clustering columns form a prefix, we do not need to fetch them for filtering, but *only* if the whole partition key is present as well. Without the partition key, clustering key prefix cannot be optimized out.

I have the solution ready, I'll post it soon.

**Comment by nyh** [comment]
Hi @psarna I'm in the middle of reviewing your patch (the github UI is *less* convenient than I thought before we started to experiment with it...), but have a general question about the entire approach or even the problem raised here:

If the full *ck* prefix is being restricted, we can do a quick read - and don't need "filtering" - on each partition. Yes, if the pk isn't restricted, we still need to go through all the partitions (so this *is* filtering), but we don't strictly need to read the entire partition and filter through it. But maybe you're going here for correctness, not absolute optimal performance?

**Comment by psarna** [comment]
@nyh yes, this issue is strictly for correctness - currently clustering restrictions can be ignored in certain corner cases, which yields incorrect results. I didn't take any optimizations involving using ck prefix for each partition into consideration (although we could create a separate issue about that).

**Comment by avikivity** [comment]
@psarna While I was able to backport to 3.1, 3.0 is much harder. Please assist.

**Comment by psarna** [comment]
I pushed a version rebased on 3.0 here: https://github.com/psarna/scylla/commits/fix_ignoring_ck_restrictions_in_filtering_for_3.0 , but I still need to compile and test it on 3.0, which I'll do soon, after dealing with #4540. After I manage to test on 3.0 properly, I'll push it to the list.

**Comment by slivne** [comment]
This was released as part of 3.0.8
